### PR TITLE
Crypto host functions

### PR DESCRIPTION
**Pull request by sisuresh**
Implements the `compute_hash_sha256` and `verify_sig_ed25519` host functions.

### CODE REVIEW

File: stellar-contract-env-host/src/host.rs
Comment (leighmcculloch) [inline stellar-contract-env-host/src/host.rs:884]: The range 0..32 is going to panic if the hash is less than 32bytes long, but not if it is longer. I don't these few lines do what you need. The try_into() is always going to succeed because there is also an into() defined for this conversion, since byte slices are always convertible into `Vec<u8>`.

If you want to check the length there's a few things you could do:

1. Simply check the length manually.
2. Convert the byte slice into a fixed-size array, and the try_into() should check length.

I think you want to get this into a `Hash` object? So you could convert to a fixed length array and put that into the hash. Something like:
```suggestion
        let hash: [u8; 32] = raw_hash
            .try_into()
            .expect("slice with incorrect length");
        Ok(self.add_host_object(Hash(hash))?.into())
```

It'd be nice if the stellar-xdr crate made this simpler and allowed you to try_into directly from a slice to a Hash, and I'll look at adding that. Hopefully it'd end up something like:
```suggestion
        let hash: Hash = raw_hash[..]
            .try_into()
            .expect("slice with incorrect length");
        Ok(self.add_host_object(hash)?.into())
```

File: stellar-contract-env-host/src/host.rs
Comment (leighmcculloch) [inline stellar-contract-env-host/src/host.rs:893]: The .clone here is unnecessary.
```suggestion
            Ok(arr)
```

File: stellar-contract-env-host/src/host.rs
Comment (leighmcculloch) [inline stellar-contract-env-host/src/host.rs:903]: The .clone here is unnecessary.

```suggestion
            Ok(arr.clone())
```

File: stellar-contract-env-host/src/host.rs
Comment (leighmcculloch) [inline stellar-contract-env-host/src/host.rs:909]: Fyi it is also common to write these type of expressions mapping the error. For example:
```suggestion
        let public_key: PublicKey =
            PublicKey::from_bytes(&key_bytes).map_err(|_| HostError::General("invalid key"))?;

```

File: stellar-contract-env-host/src/host.rs
Comment (leighmcculloch) [inline stellar-contract-env-host/src/host.rs:921]: The if block is unnecessary. For example, you can write:
```suggestion
        let verified = public_key.verify(&message_bytes[..], &sig).is_ok();
        Ok(RawVal::from_bool(verified))
```

**Review by jonjove — CHANGES_REQUESTED** [review]
We spoke offline about more than the comments here, and I gave @sisuresh some pointers to make this smaller and cleaner.

File: stellar-contract-env-host/src/host.rs
Comment (jayz22) [reply stellar-contract-env-host/src/host.rs:874]: I think @sisuresh is referring to the `.clone()` here. I believe because this is returning a result from a local reference to Vec, you need to clone otherwise the lifetime checker will complain.

Comment (jayz22) [reply stellar-contract-env-host/src/host.rs:874]: And the `visit_obj` requires that the closure takes in an immutable reference.

File: stellar-contract-env-host/src/test.rs
Comment (jayz22) [inline stellar-contract-env-host/src/test.rs:434]: I think you don't need this. Instead in your assert statement directly compare your result with `ScVal::Static`.

File: stellar-contract-env-host/src/test.rs
Comment (jayz22) [inline stellar-contract-env-host/src/test.rs:470]:
```suggestion
    assert_eq!(res, ScVal::Static(ScStatic::True));
```

File: stellar-contract-env-host/src/test.rs
Comment (jayz22) [inline stellar-contract-env-host/src/test.rs:445]: Can we get a more realistic example, like an actual transaction going through the core, and extracting its signature. This way we can be more sure this library is behaving same as the core's implementation (they should, since both are ed25519). 
And I'm not suggesting to unit test the ed25519-dalek crate, but just one example that matches with core will give us more confidence.

File: stellar-contract-env-host/src/host.rs
Comment (sisuresh) [reply stellar-contract-env-host/src/host.rs]: I removed this code

File: stellar-contract-env-host/src/host.rs
Comment (sisuresh) [reply stellar-contract-env-host/src/host.rs]: Good suggestion. Done.

File: stellar-contract-env-host/src/host.rs
Comment (sisuresh) [reply stellar-contract-env-host/src/host.rs]: I changed how all of this works

File: stellar-contract-env-host/src/host.rs
Comment (sisuresh) [reply stellar-contract-env-host/src/host.rs]: removed

File: stellar-contract-env-host/src/host.rs
Comment (sisuresh) [reply stellar-contract-env-host/src/host.rs]: removed

File: stellar-contract-env-host/src/host.rs
Comment (sisuresh) [reply stellar-contract-env-host/src/host.rs]: removed

File: stellar-contract-env-host/src/host.rs
Comment (sisuresh) [reply stellar-contract-env-host/src/host.rs]: Done

File: stellar-contract-env-host/src/host.rs
Comment (sisuresh) [reply stellar-contract-env-host/src/host.rs]: Done

File: stellar-contract-env-host/src/host.rs
Comment (sisuresh) [reply stellar-contract-env-host/src/host.rs]: Done

File: stellar-contract-env-host/src/test.rs
Comment (sisuresh) [reply stellar-contract-env-host/src/test.rs]: Good point. Done

File: stellar-contract-env-host/src/host.rs
Comment (sisuresh) [reply stellar-contract-env-host/src/host.rs]: Done

File: stellar-contract-env-host/src/test.rs
Comment (graydon) [reply stellar-contract-env-host/src/test.rs:435]: Yeah we should _definitely_ expose core crypto functions to cross-validation through the rust bridge, from rust tests (or vice versa -- but I suspect the rust test infrastructure is substantially richer and more pleasant than Catch2 and autocheck)

Comment (jayz22) [reply stellar-contract-env-host/src/test.rs:435]: Yeah that would be even better.
@graydon in that case, would we be better off using the core crypto functions directly? Is that possible?

File: stellar-contract-env-host/src/host.rs
Comment (sisuresh) [reply stellar-contract-env-host/src/host.rs:888]: Actually this should return an error if verification fails like we discussed.

File: stellar-contract-env-host/src/host.rs
Comment (jonjove) [inline stellar-contract-env-host/src/host.rs:1031]: You already have a `Result<PublicKey, HostError>`, so there's no need to do `Ok(...?)`.

File: stellar-contract-env-host/src/host.rs
Comment (jonjove) [inline stellar-contract-env-host/src/host.rs:1035]: You already have a `Result<PublicKey, HostError>`, so there's no need to do `Ok(...?)`.

File: stellar-contract-env-host/src/test.rs
Comment (jonjove) [reply stellar-contract-env-host/src/test.rs:441]: I'm not opposed to getting more realistic testing here, but I'm not going to hold up merging this to get that. Let's open an issue once this is merged to track improving the testing.

File: stellar-contract-env-host/src/host.rs
Comment (jonjove) [reply stellar-contract-env-host/src/host.rs:1031]: Copy-paste error, I meant `Result<Signature, HostError>`.

File: stellar-contract-env-host/src/host.rs
Comment (jonjove) [reply stellar-contract-env-host/src/host.rs:1035]: Copy-paste error here too, sorry.

File: stellar-contract-env-host/src/host.rs
Comment (sisuresh) [reply stellar-contract-env-host/src/host.rs]: done

File: stellar-contract-env-host/src/host.rs
Comment (sisuresh) [reply stellar-contract-env-host/src/host.rs]: done

File: stellar-contract-env-host/src/host.rs
Comment (sisuresh) [reply stellar-contract-env-host/src/host.rs]: done

**Review by jonjove — APPROVED** [review]
Looks good!

File: stellar-contract-env-host/src/test.rs
Comment (sisuresh) [reply stellar-contract-env-host/src/test.rs:441]: I have some tests on the stellar-core side that test signature verification through `create_contract` as well. I agree we should add more tests, but I don't think that should hold up merging since these PR's get out of sync with master frequently due to the parallel work being done. I'll make sure to open an issue for testing like Jon mentioned if everyone agrees.

File: stellar-contract-env-host/src/host.rs
Comment (leighmcculloch) [inline stellar-contract-env-host/src/host.rs:1020]: It looks like this returns a SCO_BINARY, but the fn should return an SCO_HASH. I don't think the hash type is added yet. cc @jayz22 who I think is adding hash in https://github.com/stellar/rs-stellar-xdr/pull/87.

File: stellar-contract-env-host/src/test.rs
Comment (leighmcculloch) [inline stellar-contract-env-host/src/test.rs:466]: Fyi you can change the signature of the test fn to be `-> Result<_, _>` and then replace the locations you want to make sure don't error with a `?`. If you prefer the panic test style, you can replace this match with a `res.expect("verification failed")`.

File: stellar-contract-env-host/src/host.rs
Comment (jayz22) [reply stellar-contract-env-host/src/host.rs:1020]: Yes that's in the process of being added. Once merged, I can update it.

Comment (jonjove) [reply stellar-contract-env-host/src/host.rs:1020]: I still don't think the hash/public key/binary thing is settled. Let's stick with binary as much as possible until we see a reason to do the other things (because doing the other things requires adding additional functions, so sticking with binary is the path of least resistance for now).

Comment (leighmcculloch) [reply stellar-contract-env-host/src/host.rs:1020]: If we're debating adding Hash/PublicKey, let's resolve that, because we're adding XDR types in https://github.com/stellar/rs-stellar-xdr/pull/87 which will lead to work to add contract spec support for them, as well as sdk types. I'd like to resolve this so we don't build all these things out if we don't need them.

Comment (jayz22) [reply stellar-contract-env-host/src/host.rs:1020]: Agree. 
This issue was previously discussed in the dev-group, in particular @graydon's comment here:
https://groups.google.com/g/stellar-dev/c/qpq4VS1ft3k/m/TUqAzXZKBQAJ explains the rationale for including these types.
I can repost in the Discord chat for more discussion.

Comment (jonjove) [reply stellar-contract-env-host/src/host.rs:1020]: Found it: https://discord.com/channels/897514728459468821/979885313377845278 (jump-cannon-dev thread "Should CONTRACT CODE remain different").

File: stellar-contract-env-host/src/test.rs
Comment (sisuresh) [reply stellar-contract-env-host/src/test.rs:466]: Thanks for tip. I'll update this in my create contract PR.

**Pull request merged**
